added support for entering/exiting critical sections (disabling inter…

UnTraDe opened the pull request:
Added support for entering and exiting critical sections either manually by calling `enter_critical` and `exit_critical` on `EspCriticalMutex`, or automatically by calling `scoped`.

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/freertos-smp.html#critical-sections

ivmarkov commented:
How is this different from the `EspMutex` implementation we already have?

UnTraDe commented:
@ivmarkov It is not just a Mutex, it's used to disable interrupts (and therefore task switching, since the FreeRTOS scheduler is also invoked with interrupt) when you need precise timing control, like bit-banging a clock signal for some sensor, etc.
It is also used when you want to modify data that is accessed both from interrupt context and the main program flow.
In ESP32 they added a Mutex as well, since it's a dual-core processor and there is no way for one core to disable interrupts for another core, so if it is used for data protection, the Mutex will also block an interrupt from the other core.

The link I refereed explains it more clearly: 

> https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/freertos-smp.html#critical-sections

ivmarkov commented:
OK so I think having an interrupt-disabling critical section utility is very useful.
(BTW: You might want to tighten your code as it does not really compile on single-core MCUs like esp32c3 and esp32s2, so you might need to put some `#[cfg(esp32c3)]` conditionals or suchlike in your code.

The other problem I have is that I'm struggling a bit where to put this code:
* `esp-idf-sys` is not a good place, because this crate is supposed to only contain raw unsafe bindings to ESP-IDF API and to help you link your binary crate against ESP-IDF
* The only exception is `EspError` (as most of those unsafe bindings are pretty difficult to use it without it), and the `no_std` panic and alloc handlers, as without those no_std binary crates do not link (of course).

And yes, according to the above rules `EspMutex` does not belong to `esp-idf-sys` either.

I think a better place for `EspMutex` and `EspCriticalSection` would be the `esp-idf-hal` crate. Are other, bare-metal MCUs putting their mutexes/critical sections in the HAL crate?

The situation is additionally complicated by `embedded-svc` on which `esp-idf-sys` (wrongly) depends, because the Mutex trait on which `EspMutex` depends is currently defined there, so as me to avoid another single-trait crate similar to the not-so-ergonomic (IMO) `&mut`-taking embedded [mutex-trait](https://github.com/rust-embedded/mutex-trait).

So if you could fix the code so that it actually does compile for esp32c3/esp32s2/esp32h2, I can eventually merge it into `esp-idf-hal`, together with moving the `EspMutex` there, and figuring out what to do with the `Mutex` trait from `embedded-svc`...

ivmarkov commented on src/include/esp-idf/bindings.h:
```diff
@@ -157,4 +157,8 @@
 #include "host/util/util.h"
 #endif
 
+#include "mdns.h"
```
Is mdns a separate ESP-IDF  component? If so, you probably need to protect it with `#ifdef` conditionals whether that component is enabled (not all ESP-IDF components are enabled by default for `cmake`-first builds and then register the component here: https://github.com/esp-rs/esp-idf-sys/blob/master/build/common.rs#L31

Also, I strongly suggest to make a separate PR for these additional includes, as I can merge them right away (provided that you address the components' issue), and they do not have to wait for the critical section thing, which is more complex.

ivmarkov commented on src/include/esp-idf/bindings.h:
```diff
@@ -157,4 +157,8 @@
 #endif
 
+#include "mdns.h"
+#include "lwip/apps/netbiosns.h"
```
Ditto - please check if this is a separate component, although I doubt that particular one.

ivmarkov commented on src/include/esp-idf/bindings.h:
```diff
@@ -157,4 +157,8 @@
 
+#include "mdns.h"
+#include "lwip/apps/netbiosns.h"
+#include "esp_smartconfig.h"
```
Ditto.

ivmarkov commented on src/interrupts.rs:
```diff
@@ -0,0 +1,41 @@
+
+// NOTE: ESP-IDF-specific
+const SPINLOCK_FREE_INIT: u32 = 0xB33FFFFF;
+pub type EspCriticalMutex = portMUX_TYPE;
```
As mentioned in the general comment, this code does not compile on single-core ESP MCUs.

ivmarkov commented on README.md:
```diff
@@ -26,9 +26,9 @@ The location where the `esp-idf` source and tools are detected and installed is
 - **`~/.platformio`**
   - This is the "standard" PlatformIO location, where all tooling as well as the ESP-IDF is installed
-- **`$ESP_IDF_SYS_PIO_INSTALL_DIR`**
+- **`$ESP_IDF_PIO_INSTALL_DIR`**
```
Obviously not necessary anymore given today's changes I did.

UnTraDe commented:
Woops, these new commits were not meant to part of this PR. Ill revert the, and add them with their Rust wrappers in a different PR.
Regarding the comment you made for the original code, I agree that it's probably more suitable for the esp-idf-hal crate.

ivmarkov commented:
> Woops, these new commits were not meant to part of this PR. Ill revert and add them with their Rust wrappers in a different PR. Regarding the comment you made for the original code, I agree that it's probably more suitable for the esp-idf-hal crate, I'll make this change and add the conditionals

I checked and unfortunately the bare metal crates in the STM32 world have a separate crate where the mutexes are kept. Not in the HAL. `cortexm::interrupts`. Anyway, if you can fix the code, I'll try to find a more appropriate place for these mutexes somehow.

UnTraDe commented:
@ivmarkov I've renamed `EspCriticalMutex` to `EspCriticalSection` like you suggested (it does sound better :)). I've also added a guard to prevent this code from being compiled and available for the single core esp32 variants. Isn't there am `esp32` conditional that I can use like: `#[cfg(esp32)]` instead of specifying `#cfg(not(...))`?

Also, I think this code can be relevant for the single-core variants as well, it just need to call the versions of `vPortEnterCritical` and `vPortExitCritical` without the mutex, since in a single-core environment it's enough to just disable interrupts to prevent anyone to access the protected resource at the same time. I just don't have the actual single-core modules to verify that it works.

ivmarkov commented:
> @ivmarkov I've renamed `EspCriticalMutex` to `EspCriticalSection` like you suggested (it does sound better :)). I've also added a guard to prevent this code from being compiled and available for the single core esp32 variants. Isn't there am `esp32` conditional that I can use like: `#[cfg(esp32)]` instead of specifying `#cfg(not(...))`?

 Of course you have these and you can use them.

> Also, I think this code can be relevant for the single-core variants as well, it just need to call the versions of `vPortEnterCritical` and `vPortExitCritical` without the mutex, since in a single-core environment it's enough to just disable interrupts to prevent anyone to access the protected resource at the same time. I just don't have the actual single-core modules to verify that it works.

Indeed. You can check something similar I did in `atomics.rs`. You can write the code and I can test it once I'm back.

ivmarkov commented on src/interrupts.rs:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    pub fn enter_critical(&mut self) {
```
Unless I'm missing something, the current implementation of `CriticalSection` cannot be used for an implementation of a "mutex". (Is this why you removed the mutex implementation?). There are two problems with it:
* `CriticalSection` does not implement `Send` and `Sync`, and as such, cannot be shared across multiple threads (say, by wrapping it in an `Arc`).
* But even if it did implement `Send` and `Sync`, the fact that `enter_critical` and `exit_critical` take a `&mut self` make it basically impossible to wrap in an `Arc`, as `Arc`s can only provide shared references, and do not provide mutable references.
* To me, in its current shape, `CriticalSection` allows only to disable interrupts for the CPU which happens to be the active one when `enter_critical`. The only mutex implementation where `CriticalSection` can be used would be of a one that is shared across threads using static mut variables, which are unfortunately unsafe themselves (besides being not so nice to use due to the fact that they are, well, static). This reminds of the limited usability of the [embedded mutex trait](https://github.com/rust-embedded/mutex-trait) which is - IMO - not very usable, and which is why we are having [our own mutex trait in embedded-svc](https://github.com/esp-rs/embedded-svc/blob/master/src/mutex.rs#L6)

Since I'm anyway in the midst of merging this code into `esp-idf-hal`, I just want to get your feedback whether I'm missing something or whether my logic is incorrect? If not, then my plan is to change `enter_critical` / `exit_critical` to take a shrared `&self` reference by wrapping the `portMUX_TYPE` in an `UnsafeCell` instance. Using `UnsafeCell` in this context should be safe (pun not intended) because `UnsafeCell::get()` (which is an unsafe call returning a `*mut PORTMUX_TYPE`) will only be used to send this reference to `vPortEnterCritical`/`vPortReleaseCritical` and the content of the `UnsafeCell` instance will not be otherwise shared with the rest of the Rust program.

ivmarkov replied:
One more *significant* complication I should've considered from the very beginning: disabling interrupts is actually very dangerous on the esp32* chips, as you have to first of all make sure that **all code** that you execute from the moment interrupts are disabled till the moment they are re-enabled again is in IRAM memory. This is done (in C) by placing an `__IRAM` attribute in front of your functions, which can be verbose and error prone, or by using special esp32 linker scripts. 

Similarly - for Rust code - you should put (I think) a `#[link_section = ".rwtext"]` attribute in front of each and every function and method that might happen to be called when interrupts are disabled. Not sure how linker scripts would work for Rust code though. The reason for this is that when interrupts are disabled, the CPU cannot load code sections from flash into IRAM, so the user has to ensure that everything which is to be executed is already in IRAM.

Considering the above, I'm still willing to merge your `CriticalSection` code (with the changes I made so that interrupt-based mutexes are supported, as well as a few other utilities), however these will be hidden behind the `--features experimental` flag, until we make sure that the whole thing with the `__IRAM` attribute really works reliably.

igrr replied:
> One more significant complication I should've considered from the very beginning: disabling interrupts is actually very dangerous on the esp32* chips, as you have to first of all make sure that all code that you execute from the moment interrupts are disabled till the moment they are re-enabled again is in IRAM memory. This is done (in C) by placing an __IRAM attribute in front of your functions, which can be verbose and error prone, or by using special esp32 linker scripts.

@ivmarkov  I don't think this is the case. There is a similar looking issue, that interrupt handlers should be placed into IRAM if you want them to allow preempting flash operations. However simply disabling interrupts doesn't harm execution of code from Flash.

ivmarkov replied:
@igrr Reading the ESP32 Technical Reference MMU section did not help me answer this simple question: is it the case, that - once configured - the MMU does NOT need any assistance from the CPU whatsoever in dealing with reads from flash? I.e. are you saying that I can have all interrupts disabled on the CPU, yet, the MMU will operate perfectly fine, handling cache misses, readings from flash and so on?

igrr replied:
That's correct, Cache/MMU hardware works with external Flash or PSRAM without relying on CPU. You can disable the interrupts and execution of code from Flash (or PSRAM, on chips which support this) will still work fine.

ivmarkov reviewed (changes requested):
Thanks for your contribution. If you can review my one (and only) comment below.

ivmarkov commented:
Your code (with some extensions) is now part of `esp-idf-hal` but hidden behind the "experimental" feature
 :) Thanks again!

weixiongmei commented:
Hi, how to use the code in Rust? Thanks~~
